LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaPayloadDecoder.cs

```diff
@@ -44,105 +60,137 @@ public async ValueTask<JObject> DecodeMessageAsync(byte[] payload, byte fport, s
-            catch
-            {
-                resultJson = JObject.Parse($"{{\"error\": \"Invalid JSON returned from '{sensorDecoder}'\", \"rawpayload\": \"{base64Payload}\"}}");
-            }
-
-            return resultJson;
         }
 
-        async Task<string> CallSensorDecoderModule(string sensorDecoderModuleUrl, byte[] payload)
+        async Task<DecodePayloadResult> CallSensorDecoderModule(string sensorDecoderModuleUrl, byte[] payload)
         {
-            var base64Payload = Convert.ToBase64String(payload);
-            string result = string.Empty;
+            var base64Payload = ((payload?.Length ?? 0) == 0) ? string.Empty : Convert.ToBase64String(payload);
 
             try
             {
-                HttpResponseMessage response = await this.decodersHttpClient.Value.GetAsync(sensorDecoderModuleUrl);
+                var httpClientToUse = this.httpClient ?? this.decodersHttpClient.Value;
+                HttpResponseMessage response = await httpClientToUse.GetAsync(sensorDecoderModuleUrl);
 
                 if (!response.IsSuccessStatusCode)
                 {
                     var badReqResult = await response.Content.ReadAsStringAsync();
 
-                    result = JsonConvert.SerializeObject(new
+                    return new DecodePayloadResult()
                     {
-                        error = $"SensorDecoderModule '{sensorDecoderModuleUrl}' returned bad request.",
-                        exceptionMessage = badReqResult ?? string.Empty,
-                        rawpayload = base64Payload
-                    });
+                        Error = $"SensorDecoderModule '{sensorDecoderModuleUrl}' returned bad request.",
+                        ErrorDetail = badReqResult ?? string.Empty,
+                    };
                 }
                 else
                 {
-                    result = await response.Content.ReadAsStringAsync();
+                    var externalRawResponse = await response.Content.ReadAsStringAsync();
+                    if (!string.IsNullOrEmpty(externalRawResponse))
+                    {
+                        try
+                        {
+                            LoRaCloudToDeviceMessage loRaCloudToDeviceMessage = null;
+                            var externalDecoderResponse = JsonConvert.DeserializeObject<Dictionary<string, object>>(externalRawResponse);
+                            if (externalDecoderResponse.TryGetValue("cloudToDeviceMessage", out var cloudToDeviceObject))
+                            {
+                                if (cloudToDeviceObject is JObject jsonObject)
+                                {
+                                    loRaCloudToDeviceMessage = jsonObject.ToObject<LoRaCloudToDeviceMessage>();
+                                }
+                                else
+                                {
+                                    loRaCloudToDeviceMessage = (LoRaCloudToDeviceMessage)cloudToDeviceObject;
```

Comment:
is this a save direct cast or should this be an indirect one?

Reply:
This was wrong, the object would never be the expected type. It will be of type `JObject` or `JValue`.